Stratis.Bitcoin/Features/BlockStore/LoopSteps/BlockStoreInnerStepFindBlocks.cs

```diff
@@ -39,43 +33,66 @@ public BlockStoreInnerStepFindBlocks(ILoggerFactory loggerFactory)
             {
-                context.BlockStoreLoop.BlockPuller.AskBlock(context.NextChainedBlock);
                 context.DownloadStack.Enqueue(context.NextChainedBlock);
-
-                if (context.DownloadStack.Count == context.BlockStoreLoop.BatchDownloadSize)
-                    context.StopFindingBlocks();
+                context.GetNextBlock();
             }
 
+            context.StopFindingBlocks();
+
+            if (context.DownloadStack.Any())
+                context.BlockStoreLoop.BlockPuller.AskForMultipleBlocks(context.DownloadStack.ToArray());
+
             this.logger.LogTrace("(-):{0}", InnerStepResult.Next);
+
             return InnerStepResult.Next;
         }
 
         private async Task<bool> ShouldStopFindingBlocks(BlockStoreInnerStepContext context)
         {
             this.logger.LogTrace("()");
 
-            bool res = (context.NextChainedBlock == null)
-                || (context.NextChainedBlock.Header.HashPrevBlock != context.InputChainedBlock.HashBlock)
-                || (context.NextChainedBlock.Height > context.BlockStoreLoop.ChainState.HighestValidatedPoW?.Height)
-                || (context.BlockStoreLoop.PendingStorage.ContainsKey(context.NextChainedBlock.HashBlock))
-                || (await context.BlockStoreLoop.BlockRepository.ExistAsync(context.NextChainedBlock.HashBlock));
+            if (context.NextChainedBlock == null)
+            {
+                this.logger.LogTrace("(-): {0} was null, return true", nameof(context.NextChainedBlock));
+                return true;
+            }
+
+            if (context.InputChainedBlock != null && (context.NextChainedBlock.Header.HashPrevBlock != context.InputChainedBlock.HashBlock))
+            {
+                this.logger.LogTrace("(-): {0} previous hash does not match {1} hash, return true.", nameof(context.NextChainedBlock), context.InputChainedBlock);
+                return true;
+            }
+
+            if (context.NextChainedBlock.Height > context.BlockStoreLoop.ChainState.HighestValidatedPoW?.Height)
+            {
+                this.logger.LogTrace("(-): {0} height was higher than {1}, return true.", nameof(context.NextChainedBlock), nameof(context.BlockStoreLoop.ChainState));
+                return true;
+            }
+
+            if (context.BlockStoreLoop.PendingStorage.ContainsKey(context.NextChainedBlock.HashBlock))
+            {
+                this.logger.LogTrace("(-): {0} contains {1}, return true", nameof(context.BlockStoreLoop.PendingStorage), nameof(context.NextChainedBlock));
+                return true;
+            }
+
+            if (await context.BlockStoreLoop.BlockRepository.ExistAsync(context.NextChainedBlock.HashBlock))
+            {
+                this.logger.LogTrace("(-): {0} contains {1}, return true", nameof(context.BlockStoreLoop.BlockRepository), nameof(context.NextChainedBlock));
+                return true;
+            }
+
+            if (context.DownloadStack.Count >= 10)
```

Comment:
could we remove the hardcoded value of `10` here and replace it with commented constant?